### PR TITLE
Rotations

### PR DESCRIPTION
**Pull request by tfamprikis**
adding functions to track orientations issue #167

### CODE REVIEW

**Comment by stefsmeets** [comment]
Hi @tfamprikis thanks for this! This looks like it will be quite a useful starting point. Do you plan on adding something more? Otherwise we will start integrating this into gemdat.

**Comment by tfamprikis** [comment]
hey, this was just a test to see how this works.

i still have a couple of issues that havent been able to fix. mainly
- I cant seem to be able to symmetrize my directions (created by gr.bond_direct() ). this is what I'm trying to do in lines 28-59 of test4.py. It works in matlab but not in python...
- I've trouble with normalizing the probability plot (in polar coordinates) by the areas of the sphere, it becomes problematic near the poles, creating artifacts.

otherwise some other functions from test4.py can also be moved to gemdat_rotation (autocorr(), autocorr_par(), a function to plot distribution of bond lenghts)

I can add the data (.xml) i've been using to test to test_data and some figures of what the symmetrization looks like when it works.

all that said, if you think you can still work with what I have please go ahead...

**Comment by stefsmeets** [comment]
Okay, no worries! The test at leas worked 😅

There is a bunch of othes stuff I still want to work on as well. I'll let you know when we get back to this.

**Comment by SCiarella** [comment]
@tfamprikis Can you share the data that you used to run `test4.py`?
Also, have you made any further progress about the rotations that is not already here?

**Comment by SCiarella** [comment]
This pull request should fulfil the following objectives:
- [x] compute the orientation vectors and their trajectory given the input expected structure
- [ ] apply the correct symmetry group to enhance the trajectory 
- [ ] produce the rotolinear plot
- [ ] measure and plot the autocorrelation

The analysis of rotational transitions will be the object to the next pull request

**Comment by SCiarella** [comment]
You can look at [this notebook](https://github.com/GEMDAT-repos/gemdat-notebooks/blob/main/rotations.ipynb) to get an overview of this addition.

**Review by stefsmeets — CHANGES_REQUESTED** [review]
Hi @SCiarella thanks for fixing up this PR! It looks like all the bits are in place, the plots look great, but I think it could use some polish before merging. I made a first pass through the code, please see my comments.

This seems to be the basic usage:

```python
bd = Orientations(trajectory, central_atoms='S', satellite_atoms='O', n_expected_neigh=8)

direction_cart = bd.get_conventional_form(normalize=False)
direction_spherical = bd.cartesian_to_spherical(direct_cart=direction_cart, degrees=True)

plots.rectilinear_plot(trajectory=direction_spherical, normalize=False)

sg = Oh_point_group()
direction_cart_sym =  bd.get_symmetric_traj(sg.sym_ops_Oh)
direction_spherical_sym = bd.cartesian_to_spherical(direct_cart=direction_cart_sym, degrees=True)

plots.rectilinear_plot(trajectory=direction_spherical_sym, normalize=False)
```

- I'm wondering if we can clean up the api a little bit, so it's a bit more inviting to use
- Have you looked at the pymatgen symmetry module? There seems to be a lot of code that can help with generating [symmetry operations for point groups](https://pymatgen.org/pymatgen.symmetry.html#pymatgen.symmetry.analyzer.PointGroupOperations). This could replace Oh_point_group with familiar pymatgen code.
- Can you also add some unit tests for the new code?
- Can you refactor the Orientations class to get rid of some of the superfluous methods(see comments)?

**Comment by stefsmeets** [comment]
We could use:

```
from pymatgen.symmetry.groups import PointGroup
g = symmetry.groups.PointGroup('m-3m')  # Oh
len(g.symmetry_ops)
# 48
g.symmetry_ops[0]
# SymmOp(affine_matrix=array([[ 0.,  0., -1.,  0.],
#       [ 0., -1.,  0.,  0.],
#       [-1.,  0.,  0.,  0.],
#       [ 0.,  0.,  0.,  1.]]))
```

File: src/gemdat/rotations.py [hunk 1]
```diff
@@ -0,0 +1,490 @@
...
+        """
+
+        frac_coord_cent, frac_coord_sat = self._fractional_coordinates()
+        combinations = self._central_satellite_matrix(distance,
+                                                      frac_coord_cent)
+        direction = frac_coord_sat[:,
+                                   combinations[:,
+                                                1], :] - frac_coord_cent[:,
+                                                                         combinations[:,
+                                                                                      0], :]
```
[the remark this answers is not in the document]
Comment (SCiarella) [reply src/gemdat/rotations.py:161]: What? I can finally read that   ( :eye_speech_bubble:-:lips:-:eye: )

File: src/gemdat/rotations.py [hunk 2]
```diff
@@ -0,0 +1,490 @@
...
+        n_ts = direction.shape[0]
+        n_bonds = direction.shape[1]
+        n_symops = sym_matrix.shape[2]
+
+        direction_sym = np.zeros((n_ts, n_bonds * n_symops, 3))
+        for m in range(n_ts):
+            for p in range(n_bonds):
+                for k in range(n_symops):
+                    direction_sym[m, p * n_symops + k, :] = np.matmul(
+                        sym_matrix[:, :, k], direction[m, p, :])
```
[the remark this answers is not in the document]
Comment (SCiarella) [reply src/gemdat/rotations.py:280]: I have solved this using `direction_sym = np.einsum('tsi,ijk->tsjk', direction, sym_matrix)` and a bit of reshaping

File: src/gemdat/plots/matplotlib/_rotations.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+    cbar = fig.colorbar(cs, label='areal probability', format='')
+
+    # Rotate the colorbar label by 180 degrees
+    cbar.ax.yaxis.set_label_coords(2.5,
+                                   0.5)  # Adjust the position of the label
+    cbar.set_label('areal probability', rotation=270, labelpad=15)
+    return fig
+
+
+def _skewed_gaussian(xgrid: np.ndarray, loc: float, scale: float,
```
[the remark this answers is not in the document]
Comment (SCiarella) [reply src/gemdat/plots/matplotlib/_rotations.py:69]: Made into a lambda function

File: src/gemdat/plots/matplotlib/_rotations.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+    hist, xedges, yedges = np.histogram2d(el, az, shape)
+
+    if normalize:
+        # Normalize by the area of the bins
+        areas = calculate_spherical_areas(shape)
+        hist = np.divide(hist, areas)
+        #replace values at the poles where normalization breaks - hacky
+        hist[0, :] = hist[1, :]
+        hist[-1, :] = hist[-2, :]
```
[the remark this answers is not in the document]
Comment (SCiarella) [reply src/gemdat/plots/matplotlib/_rotations.py:43]: agree

File: src/gemdat/plots/matplotlib/_rotations.py [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+
+    values = hist.T
+    phi = np.linspace(0, 360, np.ma.size(values, 0))
+    theta = np.linspace(0, 180, np.ma.size(values, 1))
```
[the remark this answers is not in the document]
Comment (SCiarella) [reply src/gemdat/plots/matplotlib/_rotations.py:46]: No need for the masked array. I removed it and implemented this suggestion

File: src/gemdat/plots/matplotlib/_rotations.py [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+    ax.set_xlabel(r'Bond length $[\AA]$')
+    ax.set_ylabel(r'Probability density $[\AA^{-1}]$')
+    ax.set_title('Bond Length Probability Distribution')
+    ax.legend()
+    ax.grid(True)
+
+    return fig
+
+
+def unit_vector_autocorrelation(*, trajectory: np.ndarray,
```
[the remark this answers is not in the document]
Comment (SCiarella) [reply src/gemdat/plots/matplotlib/_rotations.py:128]: Done

File: src/gemdat/plots/matplotlib/_rotations.py [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+import matplotlib.pyplot as plt
+import numpy as np
+from scipy.optimize import curve_fit
+from scipy.stats import skewnorm
+
+from gemdat.rotations import autocorrelation, calculate_spherical_areas
+
+
+def rectilinear_plot(*,
+                     trajectory: np.ndarray,
```
[the remark this answers is not in the document]
Comment (SCiarella) [reply src/gemdat/plots/matplotlib/_rotations.py:12]: done

File: src/gemdat/rotations.py [hunk 8]
```diff
@@ -0,0 +1,490 @@
...
+
+    def _fractional_coordinates(self) -> tuple[np.ndarray, np.ndarray]:
+        """Return fractional coordinates of central atoms and sattelite
+        atoms."""
+        return self._traj_cent.positions, self._traj_sat.positions
+
+    def _starting_coordinates(self) -> tuple[np.ndarray, np.ndarray]:
+        """Return starting coordinates of all central atoms and all satellite
+        atoms."""
+        return self._traj_cent.positions[1, :, :], self._traj_sat.positions[
```
[the remark this answers is not in the document]
Comment (SCiarella) [reply src/gemdat/rotations.py:74]: done

File: src/gemdat/rotations.py [hunk 9]
```diff
@@ -0,0 +1,490 @@
...
+            1, :, :]
+
+    @property
+    def _distances(self) -> np.ndarray:
+        """Calculate distances between every central atom and all satellite
+        atoms."""
+        central_start_coord, satellite_start_coord = self._starting_coordinates(
+        )
```
[the remark this answers is not in the document]
Comment (SCiarella) [reply src/gemdat/rotations.py:82]: done

File: src/gemdat/rotations.py [hunk 10]
```diff
@@ -0,0 +1,490 @@
...
+            elevation angle
+        r : float
+            radius
+        """
+        r = np.sqrt(x**2 + y**2 + z**2)
+        el = np.arcsin(z / r)
+        az = np.arctan2(y, x)
+        return az, el, r
+
+    def cartesian_to_spherical(self, direct_cart: np.ndarray,
```
[the remark this answers is not in the document]
Comment (SCiarella) [reply src/gemdat/rotations.py:329]: Numpy does not want to do that: https://github.com/numpy/numpy/issues/5228. There is an implementation on astropy https://docs.astropy.org/en/stable/api/astropy.coordinates.cartesian_to_spherical.html that would need some tweak. for the moment we are still using this implementation, which is now outside the Orientations class.

File: src/gemdat/rotations.py [hunk 11]
```diff
@@ -0,0 +1,490 @@
...
+    def _traj_cent(self) -> Trajectory:
+        """Return trajectory of center atoms."""
+        return self.traj.filter(self.center_type)
+
+    @property
+    def _traj_sat(self) -> Trajectory:
+        """Return trajectory of satellite atoms."""
+        return self.traj.filter(self.satellite_type)
+
+    def _pbc_dist(self, frac1: np.ndarray, frac2: np.ndarray) -> np.floating:
```
[the remark this answers is not in the document]
Comment (SCiarella) [reply src/gemdat/rotations.py:45]: replaced with Lattice.get_all_distances()

File: src/gemdat/rotations.py [hunk 12]
```diff
@@ -0,0 +1,490 @@
...
+        return np.array([
+            np.eye(3),
+            np.array([[-1, 0, 0], [0, 1, 0], [0, 0, 1]]),
+            np.array([[1, 0, 0], [0, -1, 0], [0, 0, 1]]),
+            np.array([[-1, 0, 0], [0, -1, 0], [0, 0, 1]]),
+            np.array([[1, 0, 0], [0, 1, 0], [0, 0, -1]]),
+            np.array([[-1, 0, 0], [0, 1, 0], [0, 0, -1]]),
+            np.array([[1, 0, 0], [0, -1, 0], [0, 0, -1]]),
+            np.array([[-1, 0, 0], [0, -1, 0], [0, 0, -1]])
+        ])
```
[the remark this answers is not in the document]
Comment (SCiarella) [reply src/gemdat/rotations.py:268]: This has been removed in favor of pymatgen.PointGroup

File: src/gemdat/rotations.py [hunk 13]
```diff
@@ -0,0 +1,490 @@
...
+            az = np.degrees(az)
+            el = np.degrees(el)
+
+        direction_spherical = np.stack((az, el, r), axis=-1)
+
+        return direction_spherical
+
+
+@dataclass
+class Oh_point_group:
```
[the remark this answers is not in the document]
Comment (SCiarella) [reply src/gemdat/rotations.py:246]: Replaced with pymatgen.PointGroup

**Comment by SCiarella** [comment]
I have cleaned the api, so now the basic usage looks like:
```
bd = Orientations(trajectory, central_atoms='S', satellite_atoms='O', n_expected_neigh=8, normalize_trajectory=False)

plots.rectilinear_plot(data=bd, symmetrize=False)

bd.set_symmetry_operations(sym_group='m-3m')

plots.rectilinear_plot(data=bd, symmetrize=True)
```
which is now using pymatgen.PointGroup to get the symmetry, so I have removed `Oh_point_group`. 
I have also refactored and removed some superfluous methods as suggested.

File: src/gemdat/rotations.py [hunk 14]
```diff
@@ -0,0 +1,335 @@
...
+
+        Returns
+        -------
+        unit_vec_traj: np.ndarray
+            Trajectories of the unit vectors
+        """
+        # Recompute also if normalized differently as expected
+        if not hasattr(self, '_unit_vectors_traj'):
+            self._compute_unit_vectors_traj()
+        return self._unit_vectors_traj
```
Comment (stefsmeets) [inline src/gemdat/rotations.py:169]: I would go with this:

```suggestion
    def _compute_unit_vectors_traj(self) -> None:
        """Computes trajectories of normalized unit vectors defined as the
        distance between a central and satellite atoms, meant to track
        orientation of molecules or clusters."""
        lattice = self.traj.lattice
        direction = self.fractional_directions(self._distances)
        unit_vec_traj = np.matmul(direction, lattice)

        if self.normalize_traj:
            unit_vec_traj = unit_vec_traj / np.linalg.norm(
                unit_vec_traj, axis=-1, keepdims=True)

        return unit_vec_traj

    def get_unit_vectors_traj(self) -> np.ndarray:
        """Returns trajectories of normalized unit vectors defined as the
        distance between a central and satellite atoms, meant to track
        orientation of molecules or clusters.

        Returns
        -------
        unit_vec_traj: np.ndarray
            Trajectories of the unit vectors
        """
        # Recompute also if normalized differently as expected
        if not hasattr(self, '_unit_vectors_traj'):
            self._unit_vectors_traj = self._compute_unit_vectors_traj()
        return self._unit_vectors_traj

```

File: src/gemdat/rotations.py [hunk 15]
```diff
@@ -0,0 +1,335 @@
...
+        Areas of the section
+    """
+    elevation_angles = np.linspace(0, 180, shape[0])
+
+    areas = np.zeros(shape, dtype=float)
+
+    for i in range(shape[1]):
+        for j in range(shape[0]):
+            azimuthal_increment = np.deg2rad(1)
+            elevation_increment = np.deg2rad(1)
```
Comment (stefsmeets) [inline src/gemdat/rotations.py:285]:
```suggestion
    azimuthal_increment = np.deg2rad(1)
    elevation_increment = np.deg2rad(1)
    
    for i in range(shape[1]):
        for j in range(shape[0]):
```

**Review by stefsmeets — APPROVED** [review]
Nice work, using the symmetries from pymatgen makes this a lot cleaner. A few small final comments from my side.

File: src/gemdat/plots/matplotlib/_rotations.py [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+                                time_units: float) -> plt.Figure:
+    """Plot the autocorrelation function of the unit vectors series.
+
+    Parameters
+    ----------
+    data : Orientations
+        The unit vector trajectories
+    time_units : float
+        The time step of the simulation in seconds, the default unit of pymatgen.trajectory.time_unit
+
```
[the remark this answers is not in the document]
Comment (SCiarella) [reply src/gemdat/plots/matplotlib/_rotations.py]: I have added a `@property` to extract this value directly from the `Orientations` class

File: src/gemdat/plots/matplotlib/_rotations.py [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+    bond_lengths = trajectory[:, :, 2].flatten()
+
+    # Plot the normalized histogram
+    hist, edges = np.histogram(bond_lengths, bins=bins, density=True)
+    bin_centers = (edges[:-1] + edges[1:]) / 2
+
+    # Fit a skewed Gaussian distribution to the data
+    params, covariance = curve_fit(
+        lambda x, a, loc, scale: skewnorm.pdf(x, a, loc, scale),
```
[the remark this answers is not in the document]
Comment (SCiarella) [reply src/gemdat/plots/matplotlib/_rotations.py]: yes indeed! thanks

**Comment by tfamprikis** [comment]
Hi guys, I've been playing with this a while. some implementations don't really make sense to me rn. should I comment here or wait for this to be merged and make issues?

**Comment by SCiarella** [comment]
@tfamprikis I think this pull request will be closed relatively soon (since it is also quite long), so it would be nice if you could open new issues to address all the problems that you see.

**Review by stefsmeets — APPROVED** [review]
LGTM 🚀

@tfamprikis I agree with @SCiarella that it's good to get this merged now and fix remaining issues later.

**Pull request merged**